File: agent/library/src/main/java/org/matilda/messages/listener/MessageListener.java
package org.matilda.messages.listener;


import org.matilda.messages.Message;
import org.matilda.messages.MessageHandlerRegistration;
import org.matilda.messages.MessageHandlerRegistry;

import javax.inject.Inject;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Predicate;


public class MessageListener {
    @Inject
    MessageHandlerRegistry mMessageHandlerRegistry;
    
    @Inject
    MessageListener() {}

    public MessageListeningInstance listen(int messageType, Predicate<Message> predicate) {
        BlockingQueue<Message> queue = new LinkedBlockingQueue<>();
        MessageHandlerRegistration registration = mMessageHandlerRegistry.registerHandler(messageType, message -> {
            if (predicate == null || predicate.test(message)) {
                queue.add(message);
            }
        });

        return new MessageListeningInstance(queue, registration);
    }

    public MessageListeningInstance listen(int messageType) {
        return listen(messageType, null);
    }
}

File: agent/library/src/main/java/org/matilda/messages/listener/MessageListener.kt
package org.matilda.messages.listener

import kotlinx.coroutines.channels.Channel
import me.tatarka.inject.annotations.Inject
import org.matilda.messages.Message
import org.matilda.messages.MessageHandlerRegistry
import java.util.function.Predicate

class MessageListener @Inject internal constructor(private val mMessageHandlerRegistry: MessageHandlerRegistry) {
    fun listen(messageType: Int, predicate: Predicate<Message>? = null): MessageListeningInstance {
        val channel = Channel<Message>()
        val registration = mMessageHandlerRegistry.registerHandler(messageType) { message: Message ->
            if (predicate == null || predicate.test(message)) {
                channel.send(message)
            }
        }
        return MessageListeningInstance(channel, registration)
    }
}
